packages/ui/aurora-grid/src/dnd/MosaicContext.tsx

```diff
@@ -0,0 +1,156 @@
+
+export type MosaicContextProviderProps = PropsWithChildren & {
+  Component?: MosaicTileComponent<any>;
+  debug?: boolean;
+};
+
+/**
+ * Root provider.
+ */
+export const MosaicContextProvider: FC<MosaicContextProviderProps> = ({
+  Component = DefaultComponent,
+  debug,
+  children,
+}) => {
+  const [delegators] = useState(
+    new Map<string, MosaicContainerProps<any>>([[DEFAULT_COMPONENT_ID, { id: DEFAULT_COMPONENT_ID, Component }]]),
+  );
+  const [activeItem, setActiveItem] = useState<MosaicDraggedItem>();
+  const [overItem, setOverItem] = useState<MosaicDraggedItem>();
+
+  const handleDragStart = (event: DragStartEvent) => {
+    setActiveItem(event.active.data.current as MosaicDraggedItem);
+  };
+
+  const handleDragOver = (event: DragOverEvent) => {
+    const { over } = event;
+    setOverItem(over?.data.current as MosaicDraggedItem);
+
+    console.log('##', over?.data.current);
+  };
+
+  const handleDragCancel = (event: DragCancelEvent) => {
+    setActiveItem(undefined);
+    setOverItem(undefined);
+  };
+
+  const handleDragEnd = (event: DragEndEvent) => {
+    if (
+      activeItem &&
+      overItem &&
+      (activeItem.container !== overItem.container || activeItem.position !== overItem.position)
+    ) {
+      console.log(activeItem, overItem);
+
+      const activeContainer = delegators.get(activeItem.container);
+      if (activeContainer) {
+        const overContainer = delegators.get(overItem.container);
+        if (overContainer && overContainer !== activeContainer) {
+          overContainer?.onMoveItem?.({
+            active: activeItem,
+            over: overItem,
+          });
+        } else {
+          activeContainer.onMoveItem?.({
+            active: activeItem,
+            over: overItem,
+          });
+        }
+      }
+    }
+
+    setActiveItem(undefined);
+    setOverItem(undefined);
+  };
+
+  const { Component: OverlayComponent = DefaultComponent } =
+    (activeItem?.container ? delegators.get(/* overItem?.container ?? */ activeItem.container) : undefined) ??
+    delegators.get(DEFAULT_COMPONENT_ID)!;
+
+  return (
+    <MosaicContext.Provider value={{ delegators, activeItem, overItem }}>
+      <DndContext
+        onDragStart={handleDragStart}
+        onDragOver={handleDragOver}
+        onDragCancel={handleDragCancel}
+        onDragEnd={handleDragEnd}
+      >
+        {debug && (
+          <Debug
+            position='bottom-left'
+            data={{
+              active: {
+                id: activeItem?.item.id,
+                container: activeItem?.container,
+              },
+              over: {
+                id: overItem?.item?.id,
+                container: overItem?.container,
+              },
+            }}
+          />
+        )}
+
+        {/* Active dragging element. */}
+        {createPortal(
+          <DragOverlay>{activeItem && <OverlayComponent data={activeItem.item} isActive={true} />}</DragOverlay>,
```

Comment:
morph based on `over` container?

Reply:
that's working now. had to battle some DOM issues.